tls/s2n_async_pkey.c

```diff
@@ -394,7 +396,17 @@ S2N_RESULT s2n_async_pkey_sign_perform(struct s2n_async_pkey_op *op, s2n_cert_pr
     RESULT_GUARD(s2n_pkey_size(pkey, &maximum_signature_length));
     RESULT_GUARD_POSIX(s2n_alloc(&sign->signature, maximum_signature_length));
 
-    RESULT_GUARD_POSIX(s2n_pkey_sign(pkey, sign->sig_alg, &sign->digest, &sign->signature));
+    /* If signature validation mode is S2N_ASYNC_PKEY_VALIDATION_STRICT
+     * then use local hash copy to sign the signature */
+    if (op->conn->config->async_pkey_validation_mode == S2N_ASYNC_PKEY_VALIDATION_STRICT) {
```

Comment:
Actually this may lead to concurrency issues. `s2n_async_pkey_sign_perform` usually is called inside the thread, while some other operations are performed on `op->connection` in a different thread (like freeing that connection), so you can't really access anything out of `op->conn` in `perform` call.

If you want to avoid redundant copy when validation mode is not strict, you can put some flag into `op` struct.

Reply:
Moved mode to op struct as well, which will get initialized while `s2n_async_pkey_decrypt_async` or `s2n_async_pkey_sign_async` call. 

For testing successful test case, we have to update mode from `S2N_ASYNC_PKEY_VALIDATION_STRICT` to `S2N_ASYNC_PKEY_VALIDATION_FAST` thus introduced an internal api to do this -> `s2n_async_pkey_op_set_validation_mode`